text/0111-feature-flags.md

```diff
@@ -0,0 +1,188 @@
+# RFC: Feature Flags
+
+## Status
+
+- Start Date: 2019-03-01
+- RFC PR: TBD
+
+## Summary
+
+This RFC defines the infrastructure pieces to support feature flags to enable experimentation and progressive development inside LWC core packages.
+
+## Motivation
+
+* It is becoming more difficult to add new features to engine and compiler due to the potential breaking hazard.
+* Some features need a period of adaptation and testing before we allow LWC users to consume them, but today the only option is to keep them in a separate branch forever.
+* The performance implications of a change is sometimes difficult to assess without hitting production servers, but at that point, it is too late since everyone has access to the same set of features today.
+* Branching code logic for certain features (e.g., slotchange event) is becoming more complicated over time, and having that in the actual code makes the code less readable.
+* Polyfilling new shadow dom semantics is almost impossible considering that we know it is going to break LWC users at some point.
+
+## Goals of this proposal
+
+* Define strategy to support incremental development of features without compromising the stability of the platform and breakages for LWC users.
+* Define ways to remove experimental/unstable code branches from production artifacts.
+* Define ways to enable and disable features at runtime.
+* Define ways to permanently enable features on production artifacts without unnecessary branching logic.
+
+## Prior Art
+
+EmberJS is probably one of the more successful frameworks when it comes to backwards-compatible changes. We can take a page from their playbook when it comes to introducing new features via Feature Flags:
+
+* https://guides.emberjs.com/release/configuring-ember/feature-flags/
+
+## Proposal
+
+New features are added within conditional statements.
+
+Code behind these flags can be conditionally enabled (or completely removed) based on your project configuration. This allows newly developed features to be selectively released when the LWC platform considers them ready for production use.
+
+### Flagging Details
+
+The flag status in the generated build is controlled by the `@lwc/features` package. This package exports a list of all available features and their current status.
+
+A feature can have one of a three flags:
+
+* `true` - The feature is present and enabled: the code behind the flag is always enabled in the generated build.
+* `null` - The feature is present but disabled in the build output. It must be enabled at runtime.
```

Comment:
@Gr8Gatsby this is the feature flag option that could work for us. the code is deployed to the browser in a disabled manner and is available to toggle at runtime